src/chainstate/stacks/miner.rs

```diff
@@ -100,39 +127,86 @@ impl<'a> StacksMicroblockBuilder<'a> {
             &MINER_BLOCK_HEADER_HASH,
         );
 
-        clarity_tx.reset_cost(initial_cost);
+        clarity_tx.reset_cost(runtime.consumed_execution.clone());
+
         Ok(StacksMicroblockBuilder {
             anchor_block,
             anchor_block_consensus_hash,
             anchor_block_height,
-            bytes_so_far,
+            runtime: runtime,
+            clarity_tx: Some(clarity_tx),
+            header_reader,
+            unconfirmed: false,
+        })
+    }
+
+    /// Create a microblock miner off of the _unconfirmed_ chaintip, i.e., resuming construction of
+    /// a microblock stream.
+    pub fn resume_unconfirmed(
+        chainstate: &'a mut StacksChainState,
+        burn_dbconn: &'a dyn BurnStateDB,
+    ) -> Result<StacksMicroblockBuilder<'a>, Error> {
+        let runtime = if let Some(ref unconfirmed_state) = chainstate.unconfirmed_state.as_ref() {
+            MicroblockMinerRuntime::from_unconfirmed_state(unconfirmed_state)
+        } else {
+            warn!("No unconfirmed state instantiated; cannot mine microblocks");
+            return Err(Error::NoSuchBlockError);
+        };
+
+        let (header_reader, _) = chainstate.reopen()?;
+        let (anchored_consensus_hash, anchored_block_hash, anchored_block_height) =
+            if let Some(ref unconfirmed) = chainstate.unconfirmed_state.as_ref() {
+                let header_info =
+                    StacksChainState::get_stacks_block_header_info_by_index_block_hash(
+                        chainstate.db(),
+                        &unconfirmed.confirmed_chain_tip,
+                    )?
+                    .ok_or(Error::NoSuchBlockError)?;
+
+                (
+                    header_info.consensus_hash,
+                    header_info.anchored_header.block_hash(),
+                    header_info.block_height,
+                )
+            } else {
+                // unconfirmed state needs to be initialized
+                return Err(Error::NoSuchBlockError)?;
+            };
+
+        let mut clarity_tx = chainstate
```

Comment:
Okay, there's an issue here that's blocking merge.

The `Drop` handler for `StacksMicroblockBuilder` does a _block_ rollback of the inner Clarity transaction.  First, I don't think this should happen, since a "block" rollback destroys all of the MARF state for the microblock stream.  The rollback should only drop the transaction state for the last microblock mined.  Second, the block-level rollback invalidates the Clarity instance inside the `UnconfirmedState`, which in turn causes a subsequent query against it via a Clarity TX to panic (since the inner MARF no longer exists).  Both of these issues need to be fixed before microblock mining is at all possible.

Reply:
Marking this as a draft until this is fixed.

Reply:
This is now fixed.